### PR TITLE
Halp: scheduler-based tests + `huh`

### PR DESCRIPTION
**Pull request by abesto**
# *DO NOT MERGE*

## What / Why

I started by trying to port https://github.com/toddsundsted/stunt/blob/300cedcdca6b3f2b64196ebebc55158f976bcce8/test/test_huh.rb. This is the first test that needs us to execute a verb, specifically to test `huh`.

## How

Created `scheduler_test_utils` to the tune of `vm_test_utils` that executes `eval` *and also commands* via a `Scheduler`. All existing tests pass with this setup; so far, so good.

An alternative could be writing a custom driver for verb execution, but since most of the verb lookup logic lives inside `Task`, that seems problematic.

## Problems / Confusion

There are many things that are surprising / confusing:

* These tests are *significantly* slower than the ones just running a VM. Am I doing it wrong? Or is this a necessary overhead of the `Scheduler`?
* The `wt` versions of tests spew logs about transaction rollback failures. Again, unsure if I'm holding it wrong, or this is a teething problem of the new storage backend.
* On some runs of `cargo test -p moor-kernel`, a random test explodes with a `OneshotReceiveError`. What... even? Is there some timing issue? Is there some shared state I can't see that shouldn't exist?
* The new test in `example.moot` fails to find a verb on the location of the player. I guess I must be holding it wrong, but I can't see how. For ease of looking, here's the test:

```
// Execute non-eval commands (i.e. verbs) with `%`
// Remember to still quote the expected output for `eval()`!
; $object = create($nothing);
; add_verb($object, {player, "xd", "accept"}, {"this", "none", "this"});
; set_verb_code($object, "accept", {"return 1;"});
; add_verb($object, {player, "xd", "hi"}, {"this", "none", "this"});
; set_verb_code($object, "hi", {
>   "return \"hello player\";"
> });
; move(player, $object);
% hi
"hello player"
```

And the result of this is a `CommandExecutionError(CommandError::NoCommandMatch)` (which I transform in the test harness to an `E_VERBNF`).

(I'm out of time for `moot` things for today, so submitting here. I suspect I'm doing something wrong on this last one, but there's also a tiny chance I guess that this is catching some actual bug)

### CODE REVIEW

File: crates/kernel/src/tasks/mod.rs [hunk 1]
```diff
@@ -162,3 +162,73 @@ pub mod vm_test_utils {
...
+        let scheduler_loop_jh = std::thread::Builder::new()
+            .name("moor-scheduler".to_string())
+            .spawn(move || loop_scheduler.run())
+            .unwrap();
+
+        let result = match subscriber.recv().unwrap() {
+            TaskWaiterResult::Error(TaskAbortedException(UncaughtException { code, .. })) => {
+                Ok(code.into())
+            }
+            TaskWaiterResult::Error(CommandExecutionError(CommandError::NoCommandMatch)) => {
```
Comment (abesto) [inline crates/kernel/src/tasks/mod.rs:201]: I feel like these transformations aren't test specific. Didn't look too hard if it exists already, but: an `Into<Var>` for `CommandExecutionError` might be neat.

File: crates/kernel/src/tasks/scheduler.rs [hunk 2]
```diff
@@ -94,15 +94,15 @@ pub enum SchedulerError {
     #[error("Could not start task (internal error)")]
     CouldNotStartTask,
     #[error("Eval compilation error")]
-    EvalCompilationError(CompileError),
+    EvalCompilationError(#[source] CompileError),
```
Comment (abesto) [inline crates/kernel/src/tasks/scheduler.rs:97]: The `#[source]` annotations give "Caused by: ..." when the error is printed.

File: crates/kernel/testsuite/moot_suite.rs [hunk 3]
```diff
@@ -50,28 +68,29 @@ impl MootState {
...
+                        command_kind: line.chars().next().unwrap().into(),
                     })
                 } else if let Some(new_player) = line.strip_prefix('@') {
                     Ok(MootState::new(session.clone(), Self::player(new_player)?))
                 } else if line.is_empty() || line.starts_with("//") {
                     Ok(self)
                 } else {
                     Err(eyre::eyre!(
-                        "Expected a command (starting `;`), a comment (starting `//`), a player switch (starting `@`), or an empty line"
+                        "Expected a command (starting `;`), a comment (starting `//`), a player switch (starting `@`), a command (starting `%`), or an empty line"
```
Comment (abesto) [inline crates/kernel/testsuite/moot_suite.rs:93]: Running out of prefixes lol. I tried `$` initially but that conflicts with `$object` as an expectation. `%` ought to work.

**Comment by rdaum** [comment]
I'm busy doing yard work and stuff today so can't dig into this too much at the moment, but just skimming:

> "These tests are significantly slower than the ones just running a VM. Am I doing it wrong? Or is this a necessary overhead of the Scheduler?
> The wt versions of tests spew logs about transaction rollback failures. Again, unsure if I'm holding it wrong, or this is a teething problem of the new storage backend."

So the first part, I'm not sure about, but maybe ties into the second. Haven't read the code yet to see how you're doing this but: The scheduler is of course asynchronous and you submit tasks and they run in parallel. If you submit a whole bunch, at once, naturally things could slow down. Likewise, if you submit a whole bunch at once you could run into transaction contention and end up with rollbacks. In *particular* the wiredtiger integration isn't fully complete in how it handles retries/rollbacks -- as I went over in my commit message, if it detects a potential conflict it raises it right away rather than at commit attempt time.  And the way the scheduler is written right now it doesn't not know how to handle that (it should immediately abort and retry, but it would have to detect that possibility for every single WorldState operation...).

I would disable WT for now since you're running into this.

The scheduler/task itself definitely has bugs, but I think just in suspend? It def used to work :-) It's also maybe a badly documented interface, so it's possible there's something in your code not right. I'll try to cherry-pick in your stuff and work through it tomorrow

It's also one of those things I'd like to rewrite. It kind of grew organically and is built in a sort of actor/CSP way with channels for coordination. I feel like I could go back now and rework it to be much simpler just using a set of atomic shared flags for coordination and maybe a couple of mutexes and queues. It kind of grew the way it did because I was formerly building all this on tokio, but I ripped that all out. 

Finally for this one:

```
; add_verb($object, {player, "xd", "hi"}, {"this", "none", "this"});
; set_verb_code($object, "hi", {
>   "return \"hello player\";"
> });
```

For this you're going to need to change the verb args from "this none this" to "none none none" or "any none none" or something similar and then make sure that $object is in the room that the player is in, or add the verb onto the player. Read up on the section on command matching in the LambdaMOO manual.. but basically "this none this" is how we reserve a verb as a non-command verb (aka an OOP "method") and commands that match have different preposition specs.

It's a bit odd.
method")

**Comment by abesto** [comment]
Flamegraph to the rescue. I'll see if I can do something about this.

flamegraph

**Comment by abesto** [comment]
Oh right. Here's what's happening:
* Task is submitted
* Loop starts, picks up and handles the task
* Loop goes to the next iteration, starts waiting for 100ms
* `scheduler.stop()`
* Loop times out, goes to the next iteration (and finds it should stop)

**Comment by abesto** [comment]
Right, the cause of the slowness is: `scheduler.stop()` is called at a point where the loop is already in `recv_timeout()`.

**Comment by abesto** [comment]
On track now, will resubmit a proper PR later.

**Pull request closed without merging**